### PR TITLE
Fix TVM backend scalar tensor inputs

### LINKED ISSUE #169188: `torch.compile(backend="tvm")` produces wrong result on simple `nn.Parameter` addition (eager = 2.0, compiled = 0.0)
### Summary

When using `torch.compile` with `backend="tvm"` on a very simple model that just adds two scalar `nn.Parameters`, eager mode returns the correct scalar 2.0, but the compiled TVM backend returns 0.0 on CUDA.

### Minimal Repro Script
```python
# -*- coding: utf-8 -*-
import torch
import torch.nn as nn

# ================== Model ==================

class MyModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.a_param = nn.Parameter(torch.tensor(1.0))
        self.b_param = nn.Parameter(torch.tensor(1.0))

    def forward(self, x):
        # x is unused, just to have an input
        return self.a_param + self.b_param


def my_model_function():
    return MyModel()


def GetInput():
    return torch.rand(1, dtype=torch.float32)


# ================== Run & Compare ==================

def main():
    # Prefer CUDA; fall back to CPU
    device = "cuda" if torch.cuda.is_available() else "cpu"
    print(f"Using device: {device}")

    model = my_model_function().to(device)
    inp = GetInput().to(device)

    # ----- Eager -----
    with torch.no_grad():
        out_eager = model(inp)

    print("=== Eager output ===")
    print(f"out_eager: {out_eager.item()}")
    print(f"  shape: {out_eager.shape}")
    print(f"  dtype: {out_eager.dtype}")
    print(f"  device: {out_eager.device}")

    # ----- torch.compile with TVM backend -----
    try:
        compiled_model = torch.compile(
            model,
            backend="tvm",
        )

        print("\n=== Running compiled_model(inp) ===")
        with torch.no_grad():
            out_compiled = compiled_model(inp)

        print("=== Compiled output ===")
        print(f"out_compiled: {out_compiled.item()}")
        print(f"  shape: {out_compiled.shape}")
        print(f"  dtype: {out_compiled.dtype}")
        print(f"  device: {out_compiled.device}")

        # Compare numerically
        try:
            close = torch.allclose(out_eager, out_compiled, rtol=1e-3, atol=1e-5)
        except Exception as e:
            close = False
            print(f"\nError during allclose comparison: {repr(e)}")

        diff = (out_compiled - out_eager).item()
        print(f"\nEager vs Compiled allclose: {close}")
        print(f"Difference (compiled - eager): {diff}")

    except Exception as e:
        print("\n=== compiled_model(inp) failed at runtime ===")
        print(f"Exception type and message: {type(e).__name__}('{e}')")


if __name__ == "__main__":
    main()
```

### Actual Output
```
Using device: cuda
=== Eager output ===
out_eager: 2.0
  shape: torch.Size([])
  dtype: torch.float32
  device: cuda:0

=== Running compiled_model(inp) ===
=== Compiled output ===
out_compiled: 0.0
  shape: torch.Size([])
  dtype: torch.float32
  device: cuda:0

Eager vs Compiled allclose: False
Difference (compiled - eager): -2.0
```

### Versions

```
PyTorch: 2.9.0 (pip wheels)

TVM: 0.18.0 (built from source)

Python: 3.10.x

CUDA / cuDNN: 12.x / 9xx (not directly used in failing path; repro uses CPU)

OS: Ubuntu 22.04

Hardware: x86_64, NVIDIA GPU available but not required to reproduce
```

cc @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184997

The TVM backend filtered runtime arguments by rank before binding them to the TVM graph executor. Rank-0 tensor inputs can represent scalar nn.Parameters captured by Dynamo, so active scalar inputs were never passed to TVM and the runtime used its default zero value instead.

Bind every runtime argument that TVM reports as an active input, independent of tensor rank, and keep skipping only arguments absent from TVM runtime metadata. Grad-requiring tensors are still detached before conversion. A narrower scalar-parameter special case would leave the same rank-based bug in place for other scalar tensor inputs, so the active-input metadata is the right source of truth.

The regression test mocks TVM's Python modules and graph executor so it can run in normal Dynamo CPU test coverage. It verifies that the active scalar runtime input reaches `set_input`, keeps its value `3.0`, is detached before conversion, and that inactive inputs are still skipped. A real TVM integration repro also confirmed the root cause: the pre-fix backend returned a wrong scalar result, while loading this PR's backend produced the eager value.

Fixes #169188
Generated by my agent

Test Plan:
- PYTHONPATH=$PWD python test/dynamo/test_backends.py TestOptimizationsCPU.test_tvm_sets_scalar_tensor_inputs_cpu
- Real TVM integration repro in an isolated Python 3.11 env with torch 2.12.1+cpu and TVM 0.15.dev118: pre-fix compiled scalar output was wrong, this PR's backend matched eager with allclose True.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased the PR and resubmitted via ghstack after adding the missing `topic: not user facing` label and strengthening the scalar-input regression test to assert the bound value. The prior pyrefly and Python 3.13 inductor-core failures failed before lint/tests ran (submodule checkout and docker-image resolution), so the fresh push should rerun CI.

**Comment by jansel** [comment]
My agent says I marked this ready for review and manually requested bobrenjc93 and williamwen42. The Auto Request Review workflow failed twice because GitHub returned a temporary `diff not_available` server-load error while fetching changed files; DrCI still reports no project failures.

**Comment by jansel** [comment]
My agent says I reran the failed Auto Request Review workflow after the GitHub diff API `not_available` infra error, and it now passes. I also requested mlazos as an additional reviewer since the PR has been waiting for review.

**Review by frgossen — CHANGES_REQUESTED** [review]
My biggest problem with this PR is that it is untested by your agent: "Real TVM integration was not run because tvm is not installed in this environment."

**Before landing, can you confirm that at least the CI tests this?** 

What Claude suggests here seems to be a good idea, especially since your agent wasnt able to run the test: also assert the value of the scalar input (e.g., self.assertEqual(scalar_input.item(), 3.0)) to confirm the right tensor was bound

nit: Redundant to mention this in the test plan: lintrunner -a

**Comment by jansel** [comment]
My agent says thanks for the review. The scalar-value assertion is already in the regression test (`self.assertEqual(3.0, scalar_input.item())`), and I amended the commit message to make the test coverage explicit: current PR CI runs the mocked Dynamo CPU regression and the standard pull/Dynamo coverage, but it does not run real TVM integration here because TVM is not installed in this environment. I also removed the redundant `lintrunner -a` entry from the Test Plan. The local targeted test passed again after rebasing; fresh PR CI is now running.

**Review by frgossen — CHANGES_REQUESTED** [review]
Someone should run real TVM integration here. If noone does or is able to, noone cares about this change.

**Comment by jansel** [comment]
My agent says I ran a real TVM integration repro for the scalar-parameter case. In an isolated Python 3.11 env with torch 2.12.1+cpu and real TVM 0.15.dev118, the existing installed TVM backend reproduced the bug (eager tensor(2.), compiled tensor(1.1771e-43), allclose False). Loading this PR's torch/_dynamo/backends/tvm.py as the backend callable in the same env produced eager tensor(2.) and compiled tensor(2.) with allclose True. I also reran PYTHONPATH=<repo> python test/dynamo/test_backends.py TestOptimizationsCPU.test_tvm_sets_scalar_tensor_inputs_cpu and lintrunner -a on the PR checkout; both passed.

**Comment by jansel** [comment]
My agent says I resubmitted the PR after updating the commit message and PR description/Test Plan to reflect the real TVM integration validation and remove the stale 'not run' line. The code diff is unchanged; the targeted Dynamo TVM regression and lintrunner -a both passed locally. Fresh CI is running.

**Review by williamwen42 — APPROVED** [review]

**Review by guan404ming — APPROVED** [review]
Thanks for the fix!

**Comment by guan404ming** [comment]
Hi @jansel, thanks for the fix here. The approach here looks right to me. This PR now conflicts with main since #189010 restructured tvm.py (relay/relax split, input info hoisted to compile time), so the diff would need a redo anyway. Would you mind if I open a new PR on top of the current structure to land this? I'd keep the same fix for the relay exec loop, and I have TVM 0.25 locally so I can verify the actual numerics end-to-end.

**Comment by jansel** [comment]
@guan404ming feel free to take this one over.  Handing off to you and closing.

**Comment by guan404ming** [comment]
Sure, thanks for that.
